ext/app/electron/WindowManager.ts

```diff
@@ -0,0 +1,95 @@
+import * as electron from "electron";
+import * as path from "path";
+import { GristLoadConfig, decodeUrl } from "app/common/gristUrls";
+import { BrowserWindow } from "electron";
+import { ElectronLoginSystem } from "./LoginSystem";
+
+export class WindowManager {
+
+  // null represents "not associated with a document".
+  private docToWindowMap: Map<string, BrowserWindow> = new Map();
+  private windowToDocMap: Map<BrowserWindow, string | null> = new Map();
+
+  constructor(
+    private gristConfig: GristLoadConfig,
+    private resolveDocId: (urlIdOrDocId: string) => Promise<string>
+  ) {}
+
+  public get(docId: string): BrowserWindow | null {
+    return this.docToWindowMap.get(docId) ?? null;
+  }
+
+  public add(docId: string | null): BrowserWindow {
+
+    if (docId) {
+      const win = this.docToWindowMap.get(docId);
+      if (win) {
+        return win;
+      }
+    }
+
+    const win = new BrowserWindow({
+      width: 1024,
+      height: 768,
+      webPreferences: {
+        nodeIntegration: false,
+        preload: path.join(__dirname, "preload.js"),
+        webviewTag: true
+      },
+      backgroundColor: "#42494B",
+      autoHideMenuBar: false,
+    });
+
+    if (docId) {
+      this.docToWindowMap.set(docId, win);
+    }
+    this.windowToDocMap.set(win, docId);
+
+    win.webContents.on("did-navigate", async (_, url) => {
+      const gristUrl = decodeUrl(this.gristConfig, new URL(url));
+      const oldDocId = this.windowToDocMap.get(win);
+      // Most of the time this would be a doc ID, but it could also be an URL ID.
+      const newDocIdOrUrlId = gristUrl.doc ?? null;
+      // oldDocId must exist because windowToDocMap knows about all existing windows.
+      this.docToWindowMap.delete(oldDocId!);
+      if (newDocIdOrUrlId) {
+        // If we are navigating to a document, we must already know about it. Otherwise the document would
+        // not even have a doc ID in the first place.
+        const newDocId = await this.resolveDocId(newDocIdOrUrlId);
+        this.docToWindowMap.set(newDocId, win);
+        this.windowToDocMap.set(win, newDocId);
+      } else {
+        this.windowToDocMap.set(win, null);
+      }
+    });
+
+    // "closed" means the window reference is already gone.
+    win.on("close", () => {
+      const oldDoc = this.windowToDocMap.get(win!);
+      this.docToWindowMap.delete(oldDoc!);
+      this.windowToDocMap.delete(win!);
+    });
```

Comment:
I feel like we could wrap these in a couple of functions, and give ourselves better consistency guarantees here :)

Reply:
I don't get what you mean - wrap what in what functions?